core/src/main/java/io/grpc/SimpleLoadBalancerFactory.java

```diff
@@ -114,25 +115,43 @@ public void handleResolvedAddresses(
         }
         addresses = newAddresses;
         nameResolutionError = null;
-        pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+        savedInterimTransport = interimTransport;
+        interimTransport = null;
+      }
+      if (savedInterimTransport != null) {
+        savedInterimTransport.closeWithRealTransports(new Supplier<T>() {
+            @Override public T get() {
+              return tm.getTransport(newAddresses);
+            }
+          });
       }
-      pendingPicksFulfillmentBatch.link(new Supplier<ListenableFuture<T>>() {
-        @Override public ListenableFuture<T> get() {
-          return tm.getTransport(newAddresses);
-        }
-      });
     }
 
     @Override
     public void handleNameResolutionError(Status error) {
-      BlankFutureProvider.FulfillmentBatch<T> pendingPicksFulfillmentBatch;
-      StatusException statusException =
-          error.augmentDescription("Name resolution failed").asException();
+      InterimTransport<T> savedInterimTransport;
+      error = error.augmentDescription("Name resolution failed");
+      synchronized (lock) {
+        savedInterimTransport = interimTransport;
+        interimTransport = null;
+        nameResolutionError = error;
+      }
+      if (savedInterimTransport != null) {
+        savedInterimTransport.closeWithError(error);
+      }
+    }
+
+    @Override
+    public void shutdown() {
+      InterimTransport<T> savedInterimTransport;
       synchronized (lock) {
-        pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
-        nameResolutionError = statusException;
+        savedInterimTransport = interimTransport;
+        interimTransport = null;
```

Comment:
Set `nameResolutionError` in order to force `pickTransport()` to return erroringTransport. Since `pickTransport()` can be called from a different thread, it is racy with shutdown so I'm pretty sure setting `nameResolutionError` is necessary.